### ISSUE TITLE
Private trait's methods reachable through a public subtrait

### ISSUE THREAD

**Issue by bluss** [rust-lang/rust#28514]
I'm not 100% sure how the privacy rules should work.

A's methods should not be callable because the trait is private, but it's reachable though the trait C that inherits A, by calling `C::a()`.

B's methods are reachable the same way, even if it's marked pub but not publically reachable.

``` rust
pub use inner::C;

mod inner {
    trait A {
        fn a(&self) { }
    }

    pub trait B {
        fn b(&self) { }
    }

    pub trait C: A + B {
        fn c(&self) { }
    }

    impl A for i32 {}
    impl B for i32 {}
    impl C for i32 {}
}

fn main() {
    // A is private
    // B is pub, not reexported
    // C : A + B is pub, reexported

    // 0.a(); // can't call
    // 0.b(); // can't call
    0.c(); // ok

    C::a(&0); // can call
    C::b(&0); // can call
    C::c(&0); // ok
}
```

[Playpen](https://play.rust-lang.org/?gist=e3157744ea2d5ae61930&version=stable)

**Comment by bluss** [comment]
cc @alexcrichton, because you've weighed in on other privacy issues

**Comment by alexcrichton** [comment]
I would personally expect a privacy error to be generated. I'm also a little surprised about how resolution allows using a method without importing the trait as well, but that may perhaps be intended behavior.

cc @rust-lang/lang, @nrc 

triage: I-nominated

**Comment by bluss** [comment]
`C::b` seems to suggest `b` is a method in the `C` trait, "inherited", but `.b()` not being callable as a method when `C` is in scope is a counterargument.

**Comment by nikomatsakis** [comment]
I think the error here is that `pub trait C: A + B` ought to require that `A` is public.

**Comment by nrc** [comment]
public or accessible from all the places `C` is? If it is in another module, it could be public but we could still have this problem (with a more complex example). The alternative feels hard to check though.

**Comment by alexcrichton** [comment]
I'd be fine so long as it was limited to "marked `pub`" because that's generally how the public/private check works in terms of what you can export.

**Comment by nikomatsakis** [comment]
I am somewhat surprised, but this test case passes too. The difference here is that the type `I32` is private, so I expect the impl to be in error.

```rust
mod inner {
    struct I32;

    trait A {
        fn a(&self) { }
    }

    pub trait B {
        fn b(&self) { }
    }

    pub trait C: A + B {
        fn c(&self) { }
    }

    impl A for I32 {}
    impl B for I32 {}
    impl C for I32 {}
}

fn main() {
}
```

**Comment by pnkfelix** [comment]
(I'm inclined to think that the invocation `C::a(&0)` from the original example is still problematic, independently of the example @nikomatsakis gave above.)

**Comment by pnkfelix** [comment]
(namely, if one changes @nikomatsakis example so that it says `pub struct I32;`, then we would still have the issue of whether 1. one can have the declaration of `pub trait C: A { ... }` and 2. whether one can call the methods from the private trait `A` on a type where only the public trait `C` is available for use.)

**Comment by nikomatsakis** [comment]
The heart of this question is what a private trait means. After some discussion in the language team, we came to the following conclusions:

- To be consistent with the rest of the rules, private types and traits should not appear in the where-clauses (or supertraits) of public traits. This would make `pub trait B: A` an error. I believe this is the in the spirit of the relevant RFC, I've not reviewed the precise wording.
- Furthermore, if we ever were to relax this rule, we would want to define when you can call a method on a private trait. It might make sense to make that illegal outside of the module that defined the trait. 

There is definitely concern that people are going to be relying on this kind of setup, though also concern that they may not be getting the guarantees that they think they are.

**Comment by nikomatsakis** [comment]
triage: P-high

**Comment by bluss** [comment]
This variation shows how using a trait bound with `C` makes all the methods from A, B, C directly callable, while just `use C` only enables the methods from `C` and not `A` nor `B`.

```rust
fn use_c<T: C>(x: &T) {
    // A is private
    // B is pub, not reexported
    // C : A + B is pub, reexported

    x.a(); // can call
    x.b(); // can call
    x.c(); // ok

    C::a(x); // can call
    C::b(x); // can call
    C::c(x); // ok
}
```

`a` is still callable both ways in nightly — is so it's not limited to just UFCS?!

**Comment by bluss** [comment]
The last comment shows a regression in nightly. `x.a()` compiles in nightly, used to be a privacy error.

**Comment by nikomatsakis** [comment]
Unfortunately, the current assignee (@aturon) is out for a while, so someone else will have to look into this.

**Comment by brson** [comment]
Is this on beta/stable yet?

**Comment by bluss** [comment]
The regression mentioned in [this comment](https://github.com/rust-lang/rust/issues/28514#issuecomment-150576772) seems to be fixed in nightly (and never reached beta). I think it was fixed in #29325

**Comment by aturon** [comment]
@petrochenkov Are you interested in taking this on?

**Comment by petrochenkov** [comment]
@aturon 
I'm very much interested in fixing the rest of privacy bugs and then laying the road to https://github.com/rust-lang/rfcs/pull/1422, but the main problem is time now. It will be clear in the next few weeks how much resources I'll be able to spend on it.

**Comment by jseyfried** [comment]
If I understand correctly, this is not an issue after #29973 given `#![deny(public_in_private)]`. Should we still disallow this with more than the `public_in_private` lint?

**Comment by jseyfried** [comment]
cc @petrochenkov

**Comment by jseyfried** [comment]
cc @nikomatsakis

**Comment by brson** [comment]
Triage: Moving this to P-medium since nobody is working on it. cc @rust-lang/lang @petrochenkov

**Comment by alexcrichton** [comment]
also ping @jseyfried, if this is [no longer an issue](https://github.com/rust-lang/rust/issues/28514#issuecomment-188730439), could you provide a summary as to why as well? It'd be great to close out!

**Comment by jseyfried** [comment]
@alexcrichton I believe this issue can only happen if a public trait has a private parent, which is already a `public_in_private` future incompatiblity warning.

**Comment by alexcrichton** [comment]
Awesome! I @rust-lang/lang, thoughts about closing? (also applying a nomination for bringing it up)

**Comment by petrochenkov** [comment]
I think this can be requalified to E-needtest and closed when https://github.com/rust-lang/rust/pull/34206 lands. I should probably add this test to https://github.com/rust-lang/rust/pull/34206 itself.

**Comment by aturon** [comment]
Switching to needstest!

**Comment by petrochenkov** [comment]
Reopening this since `private_in_public` was reverted to warn-by-default and is not going to become a hard error.

**Comment by petrochenkov** [comment]
Visibility of a trait item is inherited from its trait, so the fix is to check this trait visibility every time a trait item is named.
```rust
C::a(x); // ERROR, trait A is private / not accessible from here
```

**Comment by petrochenkov** [comment]
```rust
C::b(&0);
```
is an orthogonal problem.
[UFCS RFC](https://github.com/rust-lang/rfcs/blob/master/text/0132-ufcs.md) says that UFCS works only with traits in scope, exactly like method calls, and `B` is not in scope, so this should supposedly be a method resolution error.
```rust
C::b(&0); // ERROR no method b in the current scope
```
